lib/dash/content_protection.js

```diff
@@ -250,6 +250,179 @@ shaka.dash.ContentProtection.parseFromRepresentation = function(
+ * @property {number} recordCount
+ *   Number of records in the PlayReady object.
+ * @property {Array.<shaka.dash.ContentProtection.PlayReadyRecord>} records
+ *   Contains a variable number of records that contain license acquisition information.
+ */
+shaka.dash.ContentProtection.PlayReadyHeaderObject;
+
+/**
+ * A map of PlayReady record types to description.
+ *
+ * @const {!Object.<number, string>}
+ * @private
+ */
+var PLAYREADY_RECORD_TYPES = {
+  0x0001 : 'RIGHTS_MANAGEMENT',
+  0x0002 : 'RESERVED',
+  0x0003 : 'EMBEDDED_LICENSE'
+};
+
+/**
+ * @param {Uint16Array} recordData
+ * @param {number} recordCount
+ * @return {Array.<shaka.dash.ContentProtection.PlayReadyRecord>}
+ * @private
+ */
+shaka.dash.ContentProtection.parseRecords_ = function(recordData, recordCount) {
+  var head = 0;
+  var records = [];
+  for (var i = 0; i < recordCount; i++) {
+    var recordRaw = recordData.subarray(head);
+
+    var type = recordRaw[0];
+    var length = recordRaw[1];
+    var offset = 2;
+    var charCount = length / 2;
+    var end = charCount + offset;
+
+    // subarray end is exclusive
+    var rawValue = recordRaw.subarray(offset, end);
+    var value = String.fromCharCode.apply(null, rawValue);
+
+    records.push({
+      type: type,
+      length: length,
+      value: value
+    });
+
+    head = end;
+  }
+
+  return records;
+};
+
+/**
+ * Based off getLicenseServerURLFromInitData from dash.js
+ * https://github.com/Dash-Industry-Forum/dash.js
+ *
+ * @param {Uint16Array} bytes
+ * @return {shaka.dash.ContentProtection.PlayReadyHeaderObject}
+ * @private
+ */
+shaka.dash.ContentProtection.parsePro_ = function(bytes) {
+  var length = bytes[0] | bytes[1];
+  var recordCount = bytes[2];
+
+  var recordData = bytes.subarray(3);
+  var records = shaka.dash.ContentProtection.parseRecords_(recordData, recordCount);
+
+  return {
+    length: length,
+    recordCount: recordCount,
+    records: records
+  };
+};
+
+/**
+ * @param {!Element} xml
+ * @return {string}
+ * @private
+ */
+shaka.dash.ContentProtection.getLaurl_ = function(xml) {
+  var laurlNode = xml.getElementsByTagName('LA_URL')[0];
+  if (laurlNode) {
+    var laurl = laurlNode.childNodes[0].nodeValue;
+
+    if (laurl) {
+      return laurl;
+    }
+  }
+
+  return '';
+};
+
+/**
+ * Gets a PlayReady license URL from a content protection element
+ * containing a PlayReady Header Object
+ *
+ * @param {shaka.dash.ContentProtection.Element} element
+ * @return {string}
+ * @private
+ */
+shaka.dash.ContentProtection.getPlayReadyLicenseServerURL_ = function(element) {
+  var proNode = shaka.util.XmlUtils.findChild(element.node, 'mspr:pro');
+
+  if (!proNode) {
+    return '';
+  }
+
+  var bytes = shaka.util.Uint8ArrayUtils.fromBase64(proNode.textContent);
+  var parsedPro = shaka.dash.ContentProtection.parsePro_(new Uint16Array(bytes.buffer));
+
+  var records = parsedPro.records;
+  var parser = new DOMParser();
+  for (var i = 0; i < records.length; i++) {
```

Comment:
I would prefer it if we could avoid indexed array iteration. Can we do something like:

```
var record = records.filter(function(record) {
  return PLAYREADY_RECORD_TYPES[record.type] == 'RIGHTS_MANAGEMENT';
}).pop();

if (record) {
  var xml = ...
  return ...
} else {
  return ...
}
```